compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3129,6 +3168,131 @@ BType getTypeIntersection(BType lhsType, BType rhsType) {
+            return symTable.semanticError;
+        }
+
+        return intersectionRecord;
+    }
+
+    private BType getRestFieldIntersectionType(BType type, BMapType mapType) {
+        if (type.tag == TypeTags.RECORD) {
+            return getTypeIntersection(((BRecordType) type).restFieldType, mapType.constraint);
+        } else {
+            return getTypeIntersection(((BMapType) type).constraint, mapType.constraint);
+        }
+    }
+
+    private BErrorType createErrorType(BType lhsType, BType rhsType, BType detailType) {
+        BErrorTypeSymbol errorTypeSymbol = Symbols.createErrorSymbol(lhsType.flags, Names.EMPTY,
+                                                                     this.env.enclPkg.symbol.pkgID, null,
+                                                                     this.env.scope.owner, null, VIRTUAL);
+        errorTypeSymbol.flags |= rhsType.flags;
+        BErrorType errorType = new BErrorType(errorTypeSymbol, detailType);
+        errorType.flags |= errorTypeSymbol.flags;
+        errorTypeSymbol.type = errorType;
+        symResolver.markParameterizedType(errorType, detailType);
+        errorType.typeIdSet = BTypeIdSet.emptySet();
+
+        return errorType;
+    }
+
+    private void populateRecordFields(BRecordType recordType, BType originalType) {
+        BTypeSymbol intersectionRecordSymbol = recordType.tsymbol;
+        // If the detail type is BMapType simply ignore since the resulting detail type has `anydata` as rest type.
+        if (originalType.getKind() != TypeKind.RECORD) {
+            return;
+        }
+        BRecordType originalRecordType = (BRecordType) originalType;
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+        for (BField origField : originalRecordType.fields.values()) {
+            org.wso2.ballerinalang.compiler.util.Name origFieldName = origField.name;
+            BType recordFieldType = origField.type;
+            BVarSymbol recordFieldSymbol = new BVarSymbol(origField.symbol.flags, origFieldName,
+                                                          this.env.enclPkg.packageID, recordFieldType,
+                                                          intersectionRecordSymbol, origField.pos, SOURCE);
+            if (recordFieldType.tag == TypeTags.INVOKABLE && recordFieldType.tsymbol != null) {
+                BInvokableTypeSymbol tsymbol = (BInvokableTypeSymbol) recordFieldType.tsymbol;
+                BInvokableSymbol invokableSymbol = (BInvokableSymbol) recordFieldSymbol;
+                invokableSymbol.params = tsymbol.params;
+                invokableSymbol.restParam = tsymbol.restParam;
+                invokableSymbol.retType = tsymbol.returnType;
+                invokableSymbol.flags = tsymbol.flags;
+            }
+            String nameString = origFieldName.value;
+            fields.put(nameString, new BField(origFieldName, null, recordFieldSymbol));
+        }
+        recordType.fields.putAll(fields);
```

Comment:
Don't we have to validate/fix types for already specified fields with the same name?

Leads to issues like https://github.com/ballerina-platform/ballerina-lang/pull/26770#issuecomment-725266288

and 
```ballerina
type Detail record {
    string x;
};

type DetailTwo record {|
    int x?;
    string...;
|};

type ErrorOne error<Detail>;

type ErrorTwo error<DetailTwo>;

type IntersectionError ErrorOne & ErrorTwo;

public function main() {
    var err1 = IntersectionError("message", x = "x"); // invalid arg type in error detail field 'x', expected 'int', found 'string'
    var err2 = IntersectionError("message", x = 1); // works
}
```

Reply:
I think I should do an additional check to handle this situation. And it should be an invalid intersection in that case right?

Reply:
Yeah, it's invalid.

I think if the field is there we should check for the intersection between the already present type and this new type, and
- make it an error if there is no intersection
- if there is an intersection - set the intersection type as the field type

Reply:
Fixed